File: cadwiki-nuget/cadwiki.DllReloader/AutoCAD/AcadAssemblyUtils.vb
Option Strict On
Option Infer Off
Option Explicit On

Imports System.Reflection
Imports Autodesk.AutoCAD.Runtime

Namespace AutoCAD
    Public Class AcadAssemblyUtils
        Public Shared Function GetCommandMethodDictionarySafely(types As Type()) As _
                Dictionary(Of CommandMethodAttribute, MethodInfo)
            Dim commandMethodAttributesToMethodInfos As New Dictionary(Of CommandMethodAttribute, MethodInfo)
            For Each type As Type In types
                Dim methodInfos As MethodInfo() = type.GetMethods()

                For Each methodInfo As MethodInfo In methodInfos
                    Dim commandMethodAttributeObject As Object =
                        DoesMethodInfoHaveAutoCADCommandAttribute(methodInfo)
                    If (commandMethodAttributeObject IsNot Nothing) Then
                        Dim commandMethodAttribute As CommandMethodAttribute =
                            CType(commandMethodAttributeObject, CommandMethodAttribute)
                        commandMethodAttributesToMethodInfos.Add(commandMethodAttribute, methodInfo)
                    End If

                Next
            Next
            Return commandMethodAttributesToMethodInfos
        End Function
        Public Shared Function GetAppObjectSafely(types As Type()) As Object
            If types IsNot Nothing Then
                For Each type As Type In types
                    Dim interfaceList As List(Of Type) = type.GetInterfaces().ToList()
                    Dim isClassIExtensionApplication As Boolean = False

                    If interfaceList.Contains(GetType(IExtensionApplication)) Then
                        isClassIExtensionApplication = True
                    End If

                    If (type.IsClass And isClassIExtensionApplication) Then
                        Dim appObject As Object = Activator.CreateInstance(type)
                        Return appObject
                    End If
                Next
            End If
            Return Nothing
        End Function


        Public Shared Function GetNewestAssembly(assemblies() As Assembly, assemblyName As String, dllLocation As String) _
            As Assembly
            Dim newestAsm As Assembly = Nothing
            Dim match As Assembly = Nothing
            For Each domainAssembly As Assembly In assemblies
                Dim domainAssemblyName As AssemblyName = domainAssembly.GetName()
                If domainAssemblyName.Name.ToLower().Equals(assemblyName.ToLower()) Then
                    match = domainAssembly
                    Dim matchVersionNumber As String = GetAssemblyVersionFromFullName(domainAssembly.FullName)
                    If newestAsm Is Nothing Then
                        newestAsm = match
                    Else
                        Dim newestVersionNumber As String = GetAssemblyVersionFromFullName(newestAsm.FullName)
                        Dim comparisonResult As Integer = CompareFileVersion(matchVersionNumber, newestVersionNumber)
                        If comparisonResult = 1 Then
                            newestAsm = domainAssembly
                        End If
                    End If
                End If
            Next

            Return newestAsm
        End Function

        Public Shared Function GetAssemblyVersionFromFullName(fullName As String) As String
            Dim strArry As String() = Split(fullName, ", ")
            Dim version As String = strArry(1)
            Dim verArry As String() = Split(version, "=")
            Dim versionNumber As String = verArry(1)
            Return versionNumber
        End Function

        Public Shared Function CompareFileVersion(strFileVersion1 As String, strFileVersion2 As String) As Integer
            ' -1 = File Version 1 is less than File Version 2
            ' 0  = Versions are the same
            ' 1  = File version 1 is greater than File Version 2
            Dim intResult As Integer = 0
            Dim strAryFileVersion1() As String = Split(strFileVersion1, ".")
            Dim strAryFileVersion2() As String = Split(strFileVersion2, ".")
            Dim i As Integer
            For i = 0 To UBound(strAryFileVersion1)
                Dim num1 As Integer = CInt(strAryFileVersion1(i))
                Dim num2 As Integer = CInt(strAryFileVersion2(i))
                If num1 > num2 Then
                    intResult = 1
                ElseIf num1 < num2 Then
                    intResult = -1
                End If
                'If we have found that the result is not > or <, no need to proceed
                If intResult <> 0 Then Exit For
            Next
            Return intResult
        End Function



        Private Shared Function DoesMethodInfoHaveAutoCADCommandAttribute(methodInfo As MethodInfo) As Object
            Dim objectAttributes As Object() = methodInfo.GetCustomAttributes(True)
            For Each objectAttribute As Object In objectAttributes
                If (objectAttribute.GetType() Is GetType(CommandMethodAttribute)) Then
                    Return objectAttribute
                End If
            Next
            Return Nothing
        End Function




    End Class


End Namespace

File: cadwiki-nuget/cadwiki.DllReloader/AutoCAD/AcadAssemblyUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;

using System.Reflection;
using Autodesk.AutoCAD.Runtime;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace cadwiki.DllReloader.AutoCAD
{
    public class AcadAssemblyUtils
    {
        public static Dictionary<CommandMethodAttribute, MethodInfo> GetCommandMethodDictionarySafely(Type[] types)
        {
            var commandMethodAttributesToMethodInfos = new Dictionary<CommandMethodAttribute, MethodInfo>();
            foreach (Type @type in types)
            {
                MethodInfo[] methodInfos = type.GetMethods();

                foreach (MethodInfo methodInfo in methodInfos)
                {
                    var commandMethodAttributeObject = DoesMethodInfoHaveAutoCADCommandAttribute(methodInfo);
                    if (commandMethodAttributeObject is not null)
                    {
                        CommandMethodAttribute commandMethodAttribute = (CommandMethodAttribute)commandMethodAttributeObject;
                        commandMethodAttributesToMethodInfos.Add(commandMethodAttribute, methodInfo);
                    }

                }
            }
            return commandMethodAttributesToMethodInfos;
        }
        public static object GetAppObjectSafely(Type[] types)
        {
            if (types is not null)
            {
                foreach (Type @type in types)
                {
                    var interfaceList = @type.GetInterfaces().ToList();
                    bool isClassIExtensionApplication = false;

                    if (interfaceList.Contains(typeof(IExtensionApplication)))
                    {
                        isClassIExtensionApplication = true;
                    }

                    if (type.IsClass & isClassIExtensionApplication)
                    {
                        var appObject = Activator.CreateInstance(type);
                        return appObject;
                    }
                }
            }
            return null;
        }


        public static Assembly GetNewestAssembly(Assembly[] assemblies, string assemblyName, string dllLocation)
        {
            Assembly newestAsm = null;
            Assembly match = null;
            foreach (Assembly domainAssembly in assemblies)
            {
                var domainAssemblyName = domainAssembly.GetName();
                if (domainAssemblyName.Name.ToLower().Equals(assemblyName.ToLower()))
                {
                    match = domainAssembly;
                    string matchVersionNumber = GetAssemblyVersionFromFullName(domainAssembly.FullName);
                    if (newestAsm is null)
                    {
                        newestAsm = match;
                    }
                    else
                    {
                        string newestVersionNumber = GetAssemblyVersionFromFullName(newestAsm.FullName);
                        int comparisonResult = CompareFileVersion(matchVersionNumber, newestVersionNumber);
                        if (comparisonResult == 1)
                        {
                            newestAsm = domainAssembly;
                        }
                    }
                }
            }

            return newestAsm;
        }

        public static string GetAssemblyVersionFromFullName(string fullName)
        {
            string[] strArry = Strings.Split(fullName, ", ");
            string version = strArry[1];
            string[] verArry = Strings.Split(version, "=");
            string versionNumber = verArry[1];
            return versionNumber;
        }

        public static int CompareFileVersion(string strFileVersion1, string strFileVersion2)
        {
            // -1 = File Version 1 is less than File Version 2
            // 0  = Versions are the same
            // 1  = File version 1 is greater than File Version 2
            int intResult = 0;
            string[] strAryFileVersion1 = Strings.Split(strFileVersion1, ".");
            string[] strAryFileVersion2 = Strings.Split(strFileVersion2, ".");
            int i;
            var loopTo = Information.UBound(strAryFileVersion1);
            for (i = 0; i <= loopTo; i++)
            {
                int num1 = Conversions.ToInteger(strAryFileVersion1[i]);
                int num2 = Conversions.ToInteger(strAryFileVersion2[i]);
                if (num1 > num2)
                {
                    intResult = 1;
                }
                else if (num1 < num2)
                {
                    intResult = -1;
                }
                // If we have found that the result is not > or <, no need to proceed
                if (intResult != 0)
                    break;
            }
            return intResult;
        }



        private static object DoesMethodInfoHaveAutoCADCommandAttribute(MethodInfo methodInfo)
        {
            object[] objectAttributes = methodInfo.GetCustomAttributes(true);
            foreach (object objectAttribute in objectAttributes)
            {
                if (ReferenceEquals(objectAttribute.GetType(), typeof(CommandMethodAttribute)))
                {
                    return objectAttribute;
                }
            }
            return null;
        }




    }


}